File: src/Layout/Widgets/Widget.java

package Layout.Widgets;

import java.io.File;
import java.io.ObjectStreamException;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import javafx.scene.Node;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.thoughtworks.xstream.annotations.XStreamOmitField;

import Configuration.CompositeConfigurable;
import Configuration.Configurable;
import Configuration.IsConfig;
import Layout.Component;
import Layout.Widgets.controller.Controller;
import Layout.Widgets.controller.io.InOutput;
import Layout.Widgets.controller.io.Input;
import Layout.Widgets.controller.io.IsInput;
import Layout.Widgets.controller.io.Output;

import static Layout.Widgets.WidgetManager.WidgetSource.ANY;
import static java.lang.annotation.ElementType.TYPE;
import static java.lang.annotation.RetentionPolicy.RUNTIME;
import static util.File.FileUtil.writeFile;
import static util.functional.Util.*;

/**
 * Widget graphical component with a functionality.
 * <p>
 * The functionality is handled by widget's {@link Controller}. The controller 
 * is instantiated when widget loads. The widget-controller relationship is 1:1
 * and permanent.
 * <p>
 * Widget can be thought of as a wrapper for controller (which may be used as
 * standalone object if implementation allows). The type of widget influences
 * the lifecycle. See {@link FXMLWidget} and {@link ClassWidget}.
 * 
 * @author uranium
 */
public abstract class Widget<C extends Controller> extends Component implements CompositeConfigurable<Object> {
    
    private static final Logger LOGGER = LoggerFactory.getLogger(Widget.class);
    
    // Name of the widget. Permanent. same as factory input_name
    // it needs to be declared to support deserialization
    final String name;
    
    @XStreamOmitField private WidgetFactory factory;
    @XStreamOmitField protected C controller;
    @XStreamOmitField private Node root;
    
    // configuration
    @XStreamOmitField
    @IsConfig(name = "Is preferred", info = "Prefer this widget among all widgets of its type. If there is a request "
            + "for widget, preferred one will be selected. ")
    private boolean preferred = false;
    
    @XStreamOmitField
    @IsConfig(name = "Is ignored", info = "Ignore this widget if there is a request.")
    private boolean forbid_use = false;
    
    
    /**
     * @param {@link Widget#name}
     */
    public Widget(String name, WidgetFactory factory) {
        this.name = name;
        this.factory = factory;
    }
        
    /** {@inheritDoc} */
    @Override
    public String getName() { return name; }
    
    /**
     * Returns this widget's content.
     * <p>
     * If called the 1st time, loads this widget's content and instantiates its
     * controller. The controller will be null until the first time this method
     * is called.
     * Any subsequent call of this method will simply return the content.
     * <p>
     * {@inheritDoc}
     * @return graphical content of this widget
     */
    @Override
    public final Node load() {
        // - loads only once
        // - attaching root to the scenegraph will automatically remove it
        //   from its old location
        // - this guarantees that widget loads only once, which means:
        //   - graphics will be constructed only once
        //   - -||- controller, controller will always be in control of
        //     the correct graphics - normally we would have to load both
        //     graphics and controller multiple times because we can not
        //     assign new graphics to old controller
        // - entire state of the widget is intact with the exception of
        //   initial load at deserialisation.
        //   This also makes deserialisation the only time when configs
        //   need to be taken care of manually
        if(root==null) root = loadInitial();
        return root;
    }
    
    /**
     * Loads the widget. Solely used for {@link #load()} method called initially,
     * before the loaded content is cached. Should be called only once per life
     * cycle of the widget and internally.
     */
    protected abstract Node loadInitial();
    
    protected void initializeController() throws InstantiationException, IllegalAccessException {
        // instantiate controller
        Class cclass = getFactory().getControllerClass();
        controller = (C) cclass.newInstance();

        // inject this widget into the controller
        util.Util.setField(controller, "widget", this);

        // generate inputs
        for(Method m : cclass.getDeclaredMethods()) {
            IsInput a = m.getAnnotation(IsInput.class);
            if(a!=null) {
                int params = m.getParameterCount();
                if(Modifier.isStatic(m.getModifiers()) || params>1) 
                    throw new RuntimeException("Method " + m + " can not be an input.");

                String i_name = a.value();
                boolean isvoid = params==0; if(isvoid) System.out.println(i_name + " " + cclass);
                Class i_type = isvoid ? Void.class : m.getParameterTypes()[0];
                Consumer i_action = isvoid 
                    ?   value -> {
                            if(value!=null) 
                                throw new ClassCastException(cclass + " " + m + ": Can not cast " + value + " into Void.class");
                            try {
                                m.setAccessible(true);
                                m.invoke(controller);
                                m.setAccessible(false);
                            } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
                                LOGGER.error("Input {} in widget {} failed to process value.",i_name,name,e);
                            }
                        }
                    :   value -> {
                            try {
                                m.setAccessible(true);
                                m.invoke(controller, value);
                                m.setAccessible(false);
                            } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
                                LOGGER.error("Input {} in widget {} failed to process value.",i_name,name,e);
                            }
                        };
                controller.getInputs().create(i_name, i_type, i_action);
            }
        }
    }
    
    /**
     * Returns controller of the widget. It provides access to public behavior
     * of the widget.
     * <p>
     * The controller is instantiated when widget loads. The controller is null
     * before that and this method should not be invoked. 
     * <p>
     * Do not check the output of this method for null! Receiving null implies
     * wrong use of this method.
     * 
     * @return controller of the widget or null if widget has not been loaded
     * yet.
     */
    public C getController() {
        return controller;
    }
    
    /** @return whether this widget will be preferred over other widgets */
    public boolean isPreffered() {
        return preferred;
    }
    /** @param val whether this widget will be preferred over other widgets */
    public void setPreferred(boolean val) {
        preferred = val;
    }
    /** @return whether this widget will be ignored on widget request */
    public boolean isIgnored() {
        return forbid_use;
    }
    /** @param val whether this widget will be ignored on widget request */
    public void setIgnored(boolean val) {
        forbid_use = val;
    }
    
    /** @return factory that produces this widget */
    public WidgetFactory getFactory() {
        return factory;
    }
    
    /** @return factory information about this widget */
    public WidgetInfo getInfo() {
        return factory;
    }
    
    /** Creates a launcher for this widget with default (no predefined) settings. */
    public void exportFxwlDefault(File dir) {
        File f = new File(dir,name + ".fxwl");
        boolean ok = writeFile(f, name);
        if(!ok) LOGGER.error("Unable to export widget launcher for {} into {}", name,f);
    }
    
/******************************************************************************/

    
    /**
     * Returns whether this widget is empty.
     * <p>
     * Empty widget has no graphics. {@link Controller#isEmpty()} indicates 
     * state - that there is no data to display within widget's graphics.
     */
    public boolean isEmpty() {
        return this instanceof EmptyWidget;
    }
    
    /**
     * Returns singleton list containing the controller of this widget
     * <p>
     * {@inheritDoc}
     */
    @Override
    public Collection<Configurable<Object>> getSubConfigurable() {
        return controller==null ? listRO() : listRO(controller);
    }    
    
    /** @return input_name of the widget */
    @Override
    public String toString() {
        return name;
    }
    
    
    /**
     * @return empty widget. Use to inject fake widget instead null value.
     */
    public static Widget EMPTY() {
        return new EmptyWidget();
    }
    
/****************************** SERIALIZATION *********************************/
    
    /** Invoked just before the serialization. */
    protected Object writeReplace() throws ObjectStreamException {
        // prepare input-output bindings
        getController().getInputs().getInputs().forEach(i -> 
            properties.put("io"+i.getName(), toS(i.getSources(), (Output o) -> o.id.toString(), ":"))
        );
        // prepare configs
        Map<String,String> m = new HashMap();
        getFields().forEach(c -> m.put(c.getName(), c.getValueS()));
        properties.put("configs", m);

        return this;
    }

        
    /**
     * Invoked just after deserialization.
     * 
     * @implSpec
     * Resolve object by initializing non-deserializable fields or providing an
     * alternative instance (e.g. to adhere to singleton pattern).
     */
    protected Object readResolve() throws ObjectStreamException {
        // try to assign factory
        if (factory==null) factory = WidgetManager.getFactory(name);
        // use empty widget when no factory available
        if (factory==null) return Widget.EMPTY();
        
        // accumulate serialized inputs for later deserialiation when all widgets are ready
        properties.entrySet().stream()
                  .filter(e->e.getKey().startsWith("io"))
                  .map(e -> new IO(this,e.getKey().substring(2), (String)e.getValue()))
                  .forEach(ios::add);
        
        return this;
    }
    
    // normally we would do this in readResolve, but controller ready only after
    // widget loads
    protected void restoreConfigs() {
        if(properties.containsKey("configs")) {
            Map<String,String> m = (Map) properties.get("configs");
            m.forEach(this::setField);
            properties.remove("configs");
        }
    }
    
/******************************************************************************/
    
    
    static class IO {
        public final Widget widget;
        public final String input_name;
        public final List<Output.Id> outputs_ids = new ArrayList();

        IO(Widget widget, String name, String outputs) {
            this.widget = widget;
            this.input_name = name;
            this.outputs_ids.addAll(map(split(outputs,":",x->x),Output.Id::fromString));
        }
    }
    
    static final ArrayList<IO> ios = new ArrayList();
    public static void deserializeWidgetIO() {
        Set<Input> is = new HashSet();
        Map<Output.Id,Output> os = WidgetManager.findAll(ANY)
                     .peek((Widget w) -> w.getController().getInputs().getInputs().forEach(is::add))
                     .flatMap(w -> w.getController().getOutputs().getOutputs().stream())
                     .collect(Collectors.toMap(i->i.id, i->i));
        InOutput.inoutputs.forEach(io -> os.put(io.o.id, io.o));
        
        ios.forEach(io -> {
            Input i = io.widget.getController().getInputs().getInput(io.input_name);
            if(i!=null)
                io.outputs_ids.stream().map(os::get).filter(ISNTØ).forEach(i::bind);
        });
    }
    

    /** Widget metadata. Passed from code to program. Use on controller class. */
    @Retention(value = RUNTIME)
    @Target(value = TYPE)
    public static @interface Info {

        /**
         * Name of the widget. "" by default.
         */
        String name() default "";

        /**
         * Description of the widget.
         */
        String description() default "";

        /**
         * Version of the widget
         */
        String version() default "";

        /**
         * Author of the widget
         */
        String author() default "";

        /**
         * Main developer of the widget.
         * @return
         */
        String programmer() default "";

        /**
         * Co-developer of the widget.
         */
        String contributor() default "";

        /**
         * Last time of change.
         * @return
         */
        String year() default "";

        /**
         * How to use text.
         * <pre>
         * For example:
         * "Available actions:\n" +
         * "    Drag cover away : Removes cover\n" +
         * "    Drop image file : Adds cover\n" +
         * "    Drop audio files : Adds files to tagger\n" +
         * "    Write : Saves the tags\n" +
         * "    Open list of tagged items"
         * </pre>
         * @return
         */
        String howto() default "";

        /**
         * Any words from the author, generally about the intention behind or bugs
         * or plans for the widget or simply unrelated to anything else information.
         * <p>
         * For example: "To do: simplify user interface." or: "Discontinued."
         * @return
         */
        String notes() default "";

        /**
         * Group the widget should categorize under as. Default {@link Widget.Group.UNKNOWN}
         * @return
         */
        Widget.Group group() default Widget.Group.UNKNOWN;
    }
    
    /** Widget's intended functionality. */
    public enum Group {
        APP,
        PLAYBACK,
        PLAYLIST,
        TAGGER,
        LIBRARY,
        VISUALISATION,
        OTHER,
        DEVELOPMENT,
        UNKNOWN;
    }
}

File: src/player/main/sp/it/pl/layout/Widget.kt
package sp.it.pl.layout

import io.github.oshai.kotlinlogging.KotlinLogging
import java.io.File
import java.time.Instant
import java.util.Objects
import java.util.UUID
import javafx.beans.property.ReadOnlyBooleanWrapper
import javafx.beans.value.ChangeListener
import javafx.geometry.Insets
import javafx.scene.Node
import javafx.scene.layout.Pane
import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.annotation.AnnotationTarget.ANNOTATION_CLASS
import kotlin.annotation.AnnotationTarget.CLASS
import kotlin.system.measureTimeMillis
import org.jetbrains.annotations.Blocking
import sp.it.pl.layout.WidgetIoManager.requestWidgetIOUpdate
import sp.it.pl.layout.controller.Controller
import sp.it.pl.layout.controller.ControllerLoadError
import sp.it.pl.layout.controller.ControllerNoFactory
import sp.it.pl.layout.controller.LegacyController
import sp.it.pl.layout.controller.io.IOLayer
import sp.it.pl.layout.controller.io.Input
import sp.it.pl.main.APP
import sp.it.util.Locatable
import sp.it.util.access.readOnly
import sp.it.util.async.future.Fut
import sp.it.util.async.future.awaitFx
import sp.it.util.conf.Config
import sp.it.util.conf.ConfigDelegator
import sp.it.util.conf.ConfigValueSource
import sp.it.util.conf.Configurable
import sp.it.util.conf.Configuration
import sp.it.util.conf.EditMode
import sp.it.util.conf.c
import sp.it.util.conf.cv
import sp.it.util.conf.cvn
import sp.it.util.conf.cvro
import sp.it.util.conf.def
import sp.it.util.conf.noPersist
import sp.it.util.dev.Experimental
import sp.it.util.dev.Idempotent
import sp.it.util.dev.failIf
import sp.it.util.dev.failIfNotFxThread
import sp.it.util.file.del
import sp.it.util.file.div
import sp.it.util.file.json.JsNull
import sp.it.util.file.json.JsValue
import sp.it.util.functional.Functors.F1
import sp.it.util.functional.asIs
import sp.it.util.functional.ifNotNull
import sp.it.util.functional.orNull
import sp.it.util.functional.runTry
import sp.it.util.math.max
import sp.it.util.reactive.Disposer
import sp.it.util.reactive.Handler0
import sp.it.util.reactive.attach
import sp.it.util.reactive.on
import sp.it.util.reactive.sync1If
import sp.it.util.ui.displayed
import sp.it.util.ui.findParent
import sp.it.util.ui.isAnyParentOf
import sp.it.util.ui.onNodeDispose
import sp.it.util.ui.pseudoClassChanged
import sp.it.util.ui.removeFromParent
import sp.it.util.units.millis

/**
 * Widget graphical component with a functionality.
 *
 *
 * The functionality is handled by widget's [Controller]. The controller
 * is instantiated when widget loads. The widget-controller relationship is 1:1
 * and permanent.
 *
 *
 * Widget can be thought of as a wrapper for controller (which may be used as
 * standalone object if implementation allows). The type of widget influences
 * the lifecycle.
 */
class Widget private constructor(factory: WidgetFactory<*>, isDeserialized: Boolean, state: WidgetDb): Component(state), Configurable<Any?>, ConfigDelegator, Locatable by factory {

   /**
    * Factory that produced this widget.
    *
    * Note that in case the application creates another version of the factory (e.g., when the
    * widget source code has been modified and recompiled in runtime), even though the old factory
    * will be removed from the list of factories (substituted by the new factory), this field will
    * still point to the old factory.
    */
   val factory: WidgetFactory<*> = factory

   /**
    * The controller of the widget. It provides access to public behavior of the widget.
    *
    * The controller is instantiated on [load] and destroyed in [close].
    * Null controller means that either the [load] hasn't been called yet, or [close] has already been called.
    *
    * The controller is provided by the [factory], specifically its [WidgetFactory.create] method. However:
    * * If [load] encounters an error, an instance of [ControllerLoadError] is used.
    * * If [factory] is not available (e.g., when deserializing widget), an instance of [ControllerNoFactory] is used.
    */
   var controller: Controller? = null
      private set

   /**
    * The graphics this widget is loaded in.
    *
    * It is responsibility of the caller of the [load] to set this field right after this widget was loaded.
    * There is no restriction where widget is loaded, so this field may be null.
    *
    * This field allows widget to control its lifecycle and context from its controller.
    */
   var ui: WidgetUi? = null

   /**
    * The graphics of the widget. It provides access to public behavior of the widget.
    *
    * The graphics is instantiated on [load] and destroyed in [close].
    * Null graphics means that either the [load] hasn't been called yet, or [close] has already been called.
    *
    * The graphics is provided by the [controller], specifically its [Controller.uiRoot] method.
    */
   var graphics: Pane? = null
      private set

   init {
      properties += state.properties
      fieldsRaw += state.fields
   }

   private val configs = HashMap<String, Config<Any?>>()
   override val configurableGroupPrefix = null
   override val configurableValueSource by lazy {
      object: ConfigValueSource {

         @Suppress("UNCHECKED_CAST")
         override fun register(config: Config<*>) {
            val key = configToRawKeyMapper(config)
            configs[key] = config as Config<Any?>
         }

         override fun initialize(config: Config<*>) {
            if (config.isPersistable()) {
               val key = configToRawKeyMapper(config)
               val source = fieldsRaw
               if (source.containsKey(key))
                  config.valueAsJson = source[key] ?: JsNull
            }
         }
      }
   }

   /** Content padding or null if left up on skin to decide. */
   val padding by cvn<Insets>(null).def(name = "Padding", group = "Widget", info = "Content padding or null if left up on skin to decide`. ")

   /** Whether this container loads eagerly or lazily by user. */
   val loadTypeCfg by cv(loadType).noPersist().def(name = "Load type", group = "Widget", info = "Whether this container loads eagerly or lazily by user")

   /** Whether this container or any parent is locked. */
   val lockedCfg by cv(locked).noPersist().def(name = "Locked", group = "Widget", info = "Whether this widget is locked")

   /** Whether this container or any parent is locked. */
   val lockedUnderCfg by cvro(lockedUnder).noPersist().def(name = "Locked (under)", group = "Widget", info = "Whether this widget or any parent is locked", editable = EditMode.APP)

   /** Whether this factory will be preferred on widget `find and create` requests. */
   val preferred by cv(false).def(name = "Is preferred", group = "Widget",  info = "Prefer this widget on `find and create`. ")

   /** Whether this factory will be ignored on widget `find and create` requests. */
   val forbidUse by cv(false).def(name = "Is ignored", group = "Widget", info = "Ignore this widget on `find and create`.")

   /** Name displayed in gui. Customizable. Default is component type name. */
   val customName by cv("").def(name = "Custom name", group = "Widget", info = "Name displayed in gui. User can set his own. By default component type name.")

   override val location: File
      get() = factory.location

   override val userLocation: File
      get() = fieldsRaw["node"]?.asJsStringValue()?.let { APP.location.user.widgets/it } ?: factory.userLocation

   /** [location] as a [Config]. */
   private val locationConfig: File by c(location).def(
      name = "Directory",
      group = "widget",
      info = "Directory of this widget. Where the sources are located.",
      editable = EditMode.NONE
   )

   /** [userLocation] as a [Config]. */
   private val locationUserConfig: File by c(userLocation).def(
      name = "Directory (user data)",
      group = "widget",
      info = "Directory of this widget. Where the user data are located.",
      editable = EditMode.NONE
   )

   /** Whether this widget is active/focused. Each window has 0 or 1 active widgets. Default false. */
   private val focusedImpl = ReadOnlyBooleanWrapper(false)

   /**
    * Whether this widget is active/focused.
    *
    * Each window has 0 or 1 active widgets.
    * Focused widget may not contain actual focus (for example when it has no focus traversable content).
    * Default false.
    */
   val focused by cvro(false) { focusedImpl.readOnlyProperty.readOnly() }.def(
      name = "Focused",
      group = "widget",
      info = "Whether widget is active/focused. Each window has 0 or 1 active widgets. Focused widget may not contain actual focus (for example when it widget has no focus traversable content).",
      editable = EditMode.APP
   )

   /**
    * Whether this widget has been created from persisted state or normally by application/user.
    * May be used for sensitive initialization, like setting default values that must not override user settings.
    */
   val isDeserialized: Boolean = isDeserialized

   /** Invoked in [close], after [graphics] and [controller] are disposed. */
   val onClose = Disposer()
   private var isClosed = false
   @Experimental("Temporary") val onLoad = Handler0()

   constructor(id: UUID, factory: WidgetFactory<*>, isDeserialized: Boolean): this(factory, isDeserialized, WidgetDb(id)) {
      forceLoading = false
      customName.value = factory.name
      focused.addListener(computeFocusChangeHandler())
   }

   constructor(state: WidgetDb): this(APP.widgetManager.factories.getFactory(state.factoryId).orNone(), true, state) {
      forceLoading = false
      customName.value = state.nameUi
      focused.addListener(computeFocusChangeHandler())
      properties.entries.stream()
         .filter { e -> e.key.startsWith("io") }
         .map { e -> WidgetIo(this, e.key.substring(2), (e.value as String?)!!) }
         .forEach { e -> WidgetIoManager.ios.add(e!!) }
   }

   override val name: String
      get() = customName.value


   /** Initializes (if not yet) and returns non-null [controller] and [graphics]. */
   @Idempotent
   override fun load(): Node {
      failIfNotFxThread()
      if (graphics==null) {
         controller = controller ?: instantiateController()
         if (controller==null) {
            val c = ControllerLoadError(this)
            graphics = c.uiRoot()
            controller = c
            onLoad()
         } else {
            try {
               graphics = controller!!.uiRoot()
               val isLegacy = controller!!.javaClass.isAnnotationPresent(LegacyController::class.java)
               if (isLegacy) restoreConfigs()
               graphics!!.sceneProperty() attach { IOLayer.componentSceneChanged(this) } on onClose
               updateIO()
            } catch (e: Throwable) {
               val c = ControllerLoadError(this)
               graphics = c.uiRoot()
               controller = c
               logger.error(e) { "$logName graphics creation failed" }
            }
         }
      }
      return graphics!!
   }

   val isLoaded: Boolean
      get() = graphics!=null

   private fun instantiateController(): Controller? {
      restoreDefaultConfigs()

      val c = factory.controllerType
      logger.info { "Instantiating $logName controller" }
      return runTry { c.java.getDeclaredConstructor(Widget::class.java).newInstance(this) }
         .ifError { APP.actionStream(RuntimeException("Instantiating $logName widget controller ($c) failed", it)) }
         .orNull()
   }

   @Idempotent
   override fun close() {
      if (isClosed) return
      isClosed = true
      logger.info { "$logName closing" }
      super.close()

      controller?.let { c ->
         controller = null
         val ips = c.io.i.getInputs()
         val ops = c.io.o.getOutputs()
         ips.forEach { IOLayer.componentPutRemoved(id, it) }
         ops.forEach { IOLayer.componentPutRemoved(id, it) }
         c.close()
         ips.forEach { it.dispose() }
         ops.forEach { it.dispose() }
      }

      graphics?.onNodeDispose()
      graphics?.removeFromParent()
      graphics = null
      ui?.dispose()

      WidgetIoManager.ios.removeIf { it.widget===this }
      onClose.invoke()
   }

   /** Sets state of this widget to that of a target widget. */
   fun setStateFrom(w: Widget) {
      // if widget was loaded we store its latest state, otherwise it contains serialized pme
      if (w.controller!=null) w.storeConfigs()
      properties.clear()
      properties.putAll(w.properties)
      padding.value = w.padding.value
      preferred.value = w.preferred.value
      forbidUse.value = w.forbidUse.value
      customName.value = w.customName.value
      loadType.value = w.loadType.value
      locked.value = w.locked.value

      // if this widget is loaded we apply state, otherwise it is done when it loads
      if (controller!=null) restoreConfigs()
      properties.entries.stream()
         .filter { it.key.startsWith("io") }
         .map { WidgetIo(this, it.key.substring(2), it.value as String) }
         .forEach { WidgetIoManager.ios.add(it) }
      if (controller!=null) updateIO()
   }

   override fun focus() {
      if (isClosed) return
      if (isLoaded) graphics!!.displayed.sync1If({ it }) { controller?.focus() } on onClose
      else onLoad.attach1 { focus() }
   }

   fun focusAndTraverse() {
      if (isLoaded) {
         controller!!.focus()
         focusAndTraverseFromToRoot()
      }
   }

   private fun computeFocusChangeHandler(): ChangeListener<Boolean> = ChangeListener { _, _, nv ->
         if (isLoaded) {
            graphics?.findParent { WidgetUi.STYLECLASS in it.styleClass }
               ?.asIs<Pane?>()
               ?.pseudoClassChanged("active", nv)
         }
      }

   override fun getConfig(name: String): Config<Any?>? = configs.values.find { it.name==name } ?: controller?.getConfig(name)

   override fun getConfigs(): Collection<Config<Any?>> = configs.values.toList() + controller?.getConfigs().orEmpty()

   val fieldsRaw: MutableMap<String, JsValue?>
      get() = properties.computeIfAbsent("fields") { HashMap<String, JsValue?>() }.asIs()

   private val logName: String
      get() = "Widget(name=${customName.value}, factory=${factory.id})"

   override fun toString() = this::class.toString() + " " + factory.id

   override fun toDb(): WidgetDb {
      val isLoaded = controller!=null

      // Prepare input-outputs
      // If widget is loaded, we serialize inputs & outputs
      // Otherwise we still have the deserialized inputs/outputs leave them as they are
      if (isLoaded) {
         controller!!.io.i.getInputs().forEach { i: Input<*> ->
            properties["io" + i.name] = i.getSources().joinToString(":") { it.id.toString() }
         }
      }

      // Prepare configs
      // If widget is loaded, we serialize name:value pairs
      // Otherwise we still have the deserialized name:value pairs and leave them as they are
      if (isLoaded) {
         storeConfigs()
      }

      return WidgetDb(id, factory.id, customName.value, loadType.value, locked.value, properties - "fields" - "factory", fieldsRaw)
   }

   private fun storeConfigs() {
      // We store only Controller configs as configs of this widget should be defined in this
      // class as fields and serialized/deserialized automatically. Only Controller is created
      // manually, Widget configs should be independent and full auto...

      // 1) nothing to serialize
      // 2) serializing empty list could actually rewrite previously serialized properties,
      //    not yet restored due to widget not yet loaded
      // 3) easy optimization
      if (controller==null) return
      val configsRaw = fieldsRaw
      getConfigs().forEach { c ->
         if (c.isPersistable())
            configsRaw[configToRawKeyMapper(c)] = c.valueAsJson
      }
   }

   private fun restoreConfigs() {
      val configsRaw = fieldsRaw
      if (configsRaw.isNotEmpty()) {
         getConfigs().forEach { c ->
            if (c.isPersistable()) {
               val key = configToRawKeyMapper(c)
               if (key in configsRaw)
                  c.valueAsJson = configsRaw[key] ?: JsNull
            }
         }
         configsRaw.clear() // restoration can only ever happen once
      }
   }

   @Blocking
   fun storeDefaultConfigs() {
      failIf(!isLoaded) { "Must be loaded to export default configs" }

      val configFile = userLocation / "default.json"
      val configs = getConfigs().filter { it.isPersistable() && !Objects.deepEquals(it.value, it.defaultValue) }
      val configuration = Configuration(configToRawKeyMapper).apply { collect(configs) }
      val props = configuration.save(configFile)
      APP.widgetManager.factories.defaultConfigs[userLocation] = props
   }

   private fun restoreDefaultConfigs() {
      APP.widgetManager.factories.defaultConfigs[userLocation]?.forEach(fieldsRaw::putIfAbsent)
   }

   @Blocking
   fun clearDefaultConfigs() {
      val configFile = userLocation / "default.json"
      configFile.del()
   }

   // called when widget is loaded/closed (or rather, when inputs or outputs are created/removed)
   // we need to create i/o nodes and i/o connections
   private fun updateIO() {
      // because widget inputs can be bound to other widget outputs, and because widgets can be
      // loaded passively (then its i/o does not exist yet), we need to update all widget i/os
      // because we do not know which bind to this widget
      IOLayer.componentSceneChanged(this)
      requestWidgetIOUpdate()
   }

   /** Widget metadata. Use by annotating [Controller] class. Legacy, use [WidgetCompanion] if possible.  */
   @Retention(RUNTIME)
   @Target(ANNOTATION_CLASS, CLASS)
   annotation class Info(
      /** Name of the widget. "" by default. */
      val name: String = "",
      /** Description of the widget. */
      val description: String = "",
      /** Version of the widget */
      val version: String = "",
      /** Author of the widget */
      val author: String = "",
      /** Co-developer of the widget. */
      val contributor: String = "",
      /** Last time of change. */
      val year: String = "",
      /** Information regarding how to use widget. */
      val howto: String = "",
      /** Any words from the author. */
      val notes: String = "",
      /** Tags categorizing the widget. Default empty. */
      val tags: Array<WidgetTag> = []
   )

   enum class LoadType {
      AUTOMATIC, MANUAL
   }

   companion object {
      private val logger = KotlinLogging.logger { }

      val configToRawKeyMapper = F1 { it: Config<*> -> it.name.replace(' ', '_').lowercase() }

      val focusChangedHandler: (Node?, Boolean) -> Unit = { n, allowTraversal ->
         val window = n?.scene?.window
         if (n!=null && window!=null) {
            val widgets = APP.widgetManager.widgets.findAll(WidgetSource.OPEN).filter { it.window===window }.toList()
            widgets.find {
               it.ui?.root?.isAnyParentOf(n) ?: false
            }.ifNotNull { fw ->
               widgets.forEach { w -> if (w!==fw) w.focusedImpl.value = false }
               fw.focusedImpl.value = true
               if (allowTraversal) fw.focusAndTraverseFromToRoot()
            }
         }
      }

   }
}